rust/garbage_collector/src/garbage_collector_orchestrator_v2.rs

```diff
@@ -0,0 +1,730 @@
+            absolute_cutoff_time,
+            sysdb_client,
+            dispatcher,
+            system,
+            storage,
+            root_manager,
+            cleanup_mode,
+            result_channel: None,
+            version_files: HashMap::new(),
+            file_ref_counts: HashMap::new(),
+            versions_to_delete_output: None,
+            pending_mark_versions_at_sysdb_tasks: HashSet::new(),
+            pending_list_files_at_version_tasks: HashSet::new(),
+            num_pending_tasks: 0,
+            min_versions_to_keep,
+
+            num_files_deleted: 0,
+            num_versions_deleted: 0,
+        }
+    }
+}
+
+#[derive(Error, Debug)]
+pub enum GarbageCollectorError {
+    #[error("Panic during compaction: {0}")]
+    Panic(#[from] PanicError),
+    #[error("Error sending message through channel: {0}")]
+    Channel(#[from] ChannelError),
+    #[error("Error receiving final result: {0}")]
+    Result(#[from] RecvError),
+    #[error("{0}")]
+    Generic(#[from] Box<dyn ChromaError>),
+    #[error("The task was aborted because resources were exhausted")]
+    Aborted,
+
+    #[error("Failed to construct version graph: {0}")]
+    ConstructVersionGraph(#[from] ConstructVersionGraphError),
+    #[error("Failed to compute versions to delete: {0}")]
+    ComputeVersionsToDelete(#[from] ComputeVersionsToDeleteError),
+    #[error("Failed to mark versions at sysdb: {0}")]
+    MarkVersionsAtSysDb(#[from] MarkVersionsAtSysDbError),
+    #[error("Failed to list files at version: {0}")]
+    ListFilesAtVersion(#[from] ListFilesAtVersionError),
+    #[error("Failed to delete unused files: {0}")]
+    DeleteUnusedFiles(#[from] DeleteUnusedFilesError),
+    #[error("Failed to delete versions at sysdb: {0}")]
+    DeleteVersionsAtSysDb(#[from] DeleteVersionsAtSysDbError),
+
+    #[error("Expected version file missing for collection {0}")]
+    MissingVersionFile(CollectionUuid),
+    #[error("Invariant violation: {0}")]
+    InvariantViolation(String),
+    #[error("Could not parse UUID: {0}")]
+    UnparsableUuid(#[from] uuid::Error),
+}
+
+impl ChromaError for GarbageCollectorError {
+    fn code(&self) -> ErrorCodes {
+        ErrorCodes::Internal
+    }
+}
+
+impl<E> From<TaskError<E>> for GarbageCollectorError
+where
+    E: Into<GarbageCollectorError>,
+{
+    fn from(value: TaskError<E>) -> Self {
+        match value {
+            TaskError::Panic(e) => GarbageCollectorError::Panic(e),
+            TaskError::TaskFailed(e) => e.into(),
+            TaskError::Aborted => GarbageCollectorError::Aborted,
+        }
+    }
+}
+
+#[derive(Debug)]
+struct ConstructVersionGraphRequest;
+
+#[async_trait]
+impl Orchestrator for GarbageCollectorOrchestrator {
+    type Output = GarbageCollectorResponse;
+    type Error = GarbageCollectorError;
+
+    fn dispatcher(&self) -> ComponentHandle<Dispatcher> {
+        self.dispatcher.clone()
+    }
+
+    async fn on_start(&mut self, ctx: &ComponentContext<Self>) {
```

Comment:
Not sure what the right abstraction is - when do you override `on_start()` v/s set `initial_tasks()` and the default impl of `on_start()` takes care of creating the tasks

Reply:
I have to override `on_start()` here because `ConstructVersionGraphRequest` is not a task. I would be in favor of removing `initial_tasks()` from the interface.